src/aviary/dataset_server.py

```diff
@@ -0,0 +1,210 @@
+                self.envs[env_id] = (env, time.time())
+                return {"env_id": env_id}
+
+        @self.app.post("/reset")
+        async def reset(req: EnvRequest):
+            async with self.lock:
+                env = self._get_env(req.env_id)
+
+            with handle_exc_as_http_exc():
+                obs, tools = await env.reset()
+
+            return (
+                MessagesAdapter.dump_python(obs),
+                ToolsAdapter.dump_python(tools, exclude_none=True, by_alias=True),
+            )
+
+        @self.app.post("/step")
+        async def step(req: StepRequest):
+            async with self.lock:
+                env = self._get_env(req.env_id)
+
+            with handle_exc_as_http_exc():
+                obs, *reward_done_trunc = await env.step(req.action)
+
+            obs_serialized = MessagesAdapter.dump_python(obs)
+            return obs_serialized, *reward_done_trunc
+
+        @self.app.post("/close")
+        async def close(req: EnvRequest):
+            async with self.lock:
+                env = self._get_env(req.env_id)
+                # Even if env.close() fails, untrack this. It is more likely that
+                # the env crashed, not that we somehow lost track of it.
+                del self.envs[req.env_id]
+
+                with handle_exc_as_http_exc():
+                    await env.close()
+
+            return {"env_id": req.env_id}
+
+        @self.app.post("/close_old_envs")
+        async def close_old(req: FlushRequest):
+            now = time.time()
+
+            async def close(env_id: str, env: TEnvironment) -> str | None:
+                try:
+                    await env.close()
+                except Exception:
+                    logger.exception(f"Failed to close env {env_id}")
+                    return None
+                else:
+                    del self.envs[env_id]
+                    return env_id
+
+            to_close: list[tuple[str, TEnvironment]] = []
+            async with self.lock:
+                for env_id, (env, last_used) in list(self.envs.items()):
+                    if now - last_used > req.last_used:
+                        to_close.append((env_id, env))
+
+                closed = await asyncio.gather(*list(starmap(close, to_close)))
+
+            return {
+                "closed_env_ids": [env_id for env_id in closed if env_id is not None]
+            }
+
+        @self.app.get("/info")
+        def info():
+            try:
+                dataset_len: int | None = len(self.dataset)
+            except TypeError:
+                dataset_len = None
+            return {
+                "dataset_size": dataset_len,
+                "running_env_ids": list(self.envs.keys()),
+            }
+
+    def start(self):
+        uvicorn.run(
+            self.app,
+            host=self.host,
+            port=self.port,
+            log_level="debug",
+        )
+
+    async def astart(self):
```

Comment:
Can you add a docstring here or to `start` saying the difference?

Also, shoudl we name this `serve` over `astart`?

Reply:
Will add docstring. I like `astart`, since it's an async equivalent of `start`

Reply:
Yeah, as I understand it, we could write `start` to look more like `astart`: https://www.uvicorn.org/#config-and-server-instances , but not the other way